test/functorch/test_aotdispatch.py

```diff
@@ -704,6 +704,32 @@ def f(a):
         self.assertEqual(x_ref.grad, x_test.grad)
         self.assertEqual(x_ref_view.grad, x_test_view.grad)
 
+    def test_nested_subclasses(self):
```

Comment:
I also think we want more comprehensive testing. A few test cases I can imagine being useful:

(1) function that accepts multiple inputs/outputs, that have different nesting levels (since the bookkeeping is pretty complicated, so we should torture test it a bit). Example: `def f(wrapper(wrapper(tensor)), wrapper(tensor), tensor, wrapper(wrapper(tensor))`

(2) testing that dynamo can handle nested subclasses ok. For example, if one of the inputs to a compiled function is a nested wrapper subclass, dynamo should know to treat a getattr(nested_subclass, inner_tensor) as returning **another** wrapper tensor subclass. And dynamo should handle user code that takes that inner tensor and does things like:

(2a) directly accesses that inner tensor's subclass metadata

(2b) directly accesses that inner tensor's inner tensors

(2c) directly inlines any custom methods defiend on that inner tensor subclass.

I think it might be helpful to add a new wrapper tensor testing subclass with its own custom metadata and custom method(s), something like:
```
class CustomMethodTensor:
    elem: torch.Tensor:
    constant_int = 3
    
    def add_constant(self, other):
        return self + other + self.constant_int
```

(2a) grabs

Reply:
Another test that would be good to have is that we get a reasonable error if you have a fw out that is a nested subclass, but at runtime the gradient that comes in is a (non-nested) subclass